### DISCUSSION TITLE
SmartDate Validation/Formatting

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#4186] [Questions]
A strange error is being reported in one of our applications.
We have a SmartDate property in our business object. From the Formatting code in our UI if the user doesn't enter any date we set the SmartDate property in our object to 'new SmartDate(true)'. 
We have added a business rule to check if SmartDate is empty but when I debug the business rule's Execute method is never called. Although I could see the debugger is hitting this line.
set { SetProperty(DateOrderedProperty, value)  of our SmartDate read/write property.
Any ideas?

Kind Regards

**Comment by rockfordlhotka** [comment]
Is it possible that the existing value equals the new value?

Not that they are the same object - but I think `SetProperty` uses equality to determine whether to run rules. So if the `SmartDate` instance `.Equals` the existing value, then the property would be considered unchanged and no rules would be run.

**Reply by Chicagoan2016** [reply]
Thank you Sir. I am going to run some tests.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka The default value of SmartDate property in our object is also showing as new SmartDate(true). 
By default, I mean before we set the value in our UI.
Is this by design? 

Kind Regards

**Reply by rockfordlhotka** [reply]
SmartDate is a value type, so the property won't be null - it will be a SmartDate, so yes, this is by design.

**Reply by Chicagoan2016** [reply]
I am a little stumped : )
What should our formatting code in UI return for the validation rule to run?
Kind regards

**Reply by rockfordlhotka** [reply]
SmartDate is designed to be used behind a `string` property. The value it provides is to translate between text input and a date value.

It sounds like you are creating a property of type SmartDate and binding it to something - a date picker or something?

The intent is that your property would be like this:

```csharp
    public static readonly PropertyInfo<SmartDate> MyDateProperty = RegisterProperty<SmartDate>(nameof(MyDate), null, new SmartDate());
    public string MyDate
    {
      get => GetPropertyConvert<SmartDate, string>(MyDateProperty);
      set => SetPropertyConvert<SmartDate, string>(MyDateProperty, value);
    }
```

**Reply by Chicagoan2016** [reply]
Yes, you are right, we are indeed using SmartDate Property in our business object.
If memory serves, we have tried using a separate string property but with free-form data entry we were getting exceptions from SetPropertyConvert if, for instance,  a user would mistype 05/05/2005 as 55/05/2005. 

We couldn't figure out where to handle exceptions from SetProperytConvert.

Kind Regards
